cmd/tailscale/cli/whois.go

```diff
@@ -0,0 +1,90 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+package cli
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"flag"
+	"fmt"
+	"os"
+	"strings"
+	"text/tabwriter"
+
+	"github.com/peterbourgon/ff/v3/ffcli"
+)
+
+var whoisCmd = &ffcli.Command{
+	Name:       "whois",
+	ShortUsage: "whois [--json] [ip|ip:port]",
+	ShortHelp:  "Show the machine and user associated with a Tailscale IP",
+	LongHelp: strings.TrimSpace(`
+	'tailscale whois' shows the machine and user associated with a Tailscale IP.
+	`),
+	UsageFunc: usageFunc,
+	Exec:      runWhoIs,
+	FlagSet: func() *flag.FlagSet {
+		fs := newFlagSet("whois")
+		fs.BoolVar(&whoIsArgs.json, "json", false, "output in JSON format")
+		return fs
+	}(),
+}
+
+var whoIsArgs struct {
+	json bool
+}
+
+func runWhoIs(ctx context.Context, args []string) error {
+	if len(args) > 1 {
+		fmt.Println(args)
+		return errors.New("too many arguments, expected at most one peer")
+	} else if len(args) == 0 {
+		return errors.New("missing argument, expected one peer")
+	}
+	who, err := localClient.WhoIs(ctx, args[0])
+	if err != nil {
+		return err
+	}
+	if whoIsArgs.json {
+		ec := json.NewEncoder(Stdout)
+		ec.SetIndent("", "  ")
+		ec.Encode(who)
+		return nil
+	}
+
+	w := tabwriter.NewWriter(os.Stdout, 10, 5, 5, ' ', 0)
+	fmt.Fprintf(w, "Machine:\n")
+	fmt.Fprintf(w, "  Name:\t%s\n", strings.TrimSuffix(who.Node.Name, "."))
+	fmt.Fprintf(w, "  ID:\t%s\n", who.Node.StableID)
+	if who.Node.IsTagged() {
+		fmt.Fprintf(w, "  Tags:\t%s\n", strings.Join(who.Node.Tags, ", "))
+	}
+	fmt.Fprintf(w, "  Addresses:\t%s\n", who.Node.Addresses)
+	if len(who.Node.AllowedIPs) > 2 {
+		fmt.Fprintf(w, "  AllowedIPs:\t%s\n", who.Node.AllowedIPs[2:])
+	}
+	if !who.Node.IsTagged() {
+		fmt.Fprintln(w, "User:")
+		fmt.Fprintf(w, "  Name:\t%s\n", who.UserProfile.LoginName)
+		fmt.Fprintf(w, "  ID:\t%d\n", who.UserProfile.ID)
+	}
+	w.Flush()
+	w = nil // avoid accidental use
+
+	if cm := who.CapMap; len(cm) > 0 {
```

Comment:
Why not using the tabwriter for this block?

Reply:
oh this was a pain to get right between all the various layers of format coming down from control. Added a comment